### PR TITLE
Improve inherited annotations

### PR DESCRIPTION
**Pull request by hughcapet**
- Annotate PVC on Sync/Update, not only change PVC template
- Don't rotate pods when only annotations changed
- Annotate Logical Backup's and Pooler's pods
- Annotate PDB, Endpoints created by the Operator, Secrets, Logical Backup jobs

Inherited annotations are only added/updated, not removed

addresses #2642

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -513,9 +519,9 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *appsv1.StatefulSet) *compa
 				reasons = append(reasons, fmt.Sprintf("new statefulset's name for volume %d does not match the current one", i))
 				continue
 			}
-			if !reflect.DeepEqual(c.Statefulset.Spec.VolumeClaimTemplates[i].Annotations, statefulSet.Spec.VolumeClaimTemplates[i].Annotations) {
+			if changed, reason := c.compareAnnotations(c.Statefulset.Spec.VolumeClaimTemplates[i].Annotations, statefulSet.Spec.VolumeClaimTemplates[i].Annotations); changed {
 				needsReplace = true
-				reasons = append(reasons, fmt.Sprintf("new statefulset's annotations for volume %q does not match the current one", name))
+				reasons = append(reasons, fmt.Sprintf("new statefulset's annotations for volume %q does not match the current one: ", name)+reason)
```
Comment (FxKu) [inline pkg/cluster/cluster.go:524]:
```suggestion
				reasons = append(reasons, fmt.Sprintf("new statefulset's annotations for volume %q does not match the current one: %s", name, reason))
```

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -818,6 +826,17 @@ func (c *Cluster) compareLogicalBackupJob(cur, new *batchv1.CronJob) (match bool
 	return true, ""
 }
 
+func (c *Cluster) ComparePodDisruptionBudget(cur, new *apipolicyv1.PodDisruptionBudget) (bool, string) {
```
Comment (FxKu) [inline pkg/cluster/cluster.go:829]:
```suggestion
func (c *Cluster) comparePodDisruptionBudget(cur, new *apipolicyv1.PodDisruptionBudget) (bool, string) {
```

we don't need to export it for other packages I think

File: e2e/tests/test_e2e.py [hunk 3]
```diff
@@ -909,25 +909,35 @@ def test_ignored_annotations(self):
...
             old_sts_creation_timestamp = sts.metadata.creation_timestamp
             k8s.api.apps_v1.patch_namespaced_stateful_set(sts.metadata.name, sts.metadata.namespace, annotation_patch)
-            svc = k8s.api.core_v1.read_namespaced_service('acid-minimal-cluster', 'default')
             old_svc_creation_timestamp = svc.metadata.creation_timestamp
             k8s.api.core_v1.patch_namespaced_service(svc.metadata.name, svc.metadata.namespace, annotation_patch)
 
             patch_config_ignored_annotations = {
                 "data": {
-                    "ignored_annotations": "k8s-status",
+                    "ignored_annotations": "k8s-status, foo",
```
Comment (FxKu) [inline e2e/tests/test_e2e.py:940]: this other annotation is just to test with two values?

File: pkg/cluster/sync.go [hunk 4]
```diff
@@ -309,7 +317,7 @@ func (c *Cluster) syncPodDisruptionBudget(isUpdate bool) error {
 	return nil
 }
 
-func (c *Cluster) syncStatefulSet() error {
+func (c *Cluster) syncStatefulSet(force bool) error {
```
Comment (FxKu) [inline pkg/cluster/sync.go:320]: is there way we move all checks done outside to set the value of `force` inside syncStatefulSet? It's called in so many places and unit test and I find the extra argument kinda ugly.


In general, I like that we call syncStatefulSet without extra conditions and resolve it inside the function.

File: pkg/cluster/volumes.go [hunk 5]
```diff
@@ -44,16 +44,20 @@ func (c *Cluster) syncVolumes() error {
...
 
 		// resize pvc to adjust filesystem size until better K8s support
-		if err = c.syncVolumeClaims(); err != nil {
+		if err = c.syncVolumeClaims(false); err != nil {
 			err = fmt.Errorf("could not sync persistent volume claims: %v", err)
 			return err
 		}
 	} else if c.OpConfig.StorageResizeMode == "pvc" {
-		if err = c.syncVolumeClaims(); err != nil {
+		if err = c.syncVolumeClaims(false); err != nil {
```
Comment (FxKu) [inline pkg/cluster/volumes.go:52]: actually I wonder if we should not call c.syncVolumneClaims everytime without all the ifs here but do it inside the function. But I see that for ebs and off mode we need some extra pre and post treatment here.

File: pkg/cluster/volumes.go [hunk 6]
```diff
@@ -184,21 +191,53 @@ func (c *Cluster) populateVolumeMetaData() error {
 }
 
 // syncVolumeClaims reads all persistent volume claims and checks that their size matches the one declared in the statefulset.
-func (c *Cluster) syncVolumeClaims() error {
+func (c *Cluster) syncVolumeClaims(noResize bool) error {
```
Comment (FxKu) [inline pkg/cluster/volumes.go:194]: I tend to call these negative variable ignoreSomething. I my head it's easier to process the following code, e.g. I need a few seconds more on !noResize

File: e2e/tests/test_e2e.py (hunk 3 above)
Comment (hughcapet) [reply e2e/tests/test_e2e.py:940]: oh. i didn't check that pod is deleted unconditionally in the `update_config()`. so no need to introduce changes, I guess

**Review by idanovinda — APPROVED** [review]
LGTM, had a few questions and confirmed them with Polina 👍

File: pkg/cluster/volumes.go [hunk 7]
```diff
@@ -187,18 +181,59 @@ func (c *Cluster) populateVolumeMetaData() error {
...
 	c.setProcessName("syncing volume claims")
 
-	needsResizing, err := c.volumeClaimsNeedResizing(c.Spec.Volume)
+	ignoreResize := false
+
+	if c.OpConfig.StorageResizeMode == "off" || c.OpConfig.StorageResizeMode == "ebs" {
+		ignoreResize = true
+		if c.OpConfig.StorageResizeMode == "off" {
+			c.logger.Infof("Storage resize is disabled (storage_resize_mode is off). Skipping volume size sync.")
+		}
```
Comment (FxKu) [inline pkg/cluster/volumes.go:190]:
```suggestion
		c.logger.Debugf("Storage resize mode is set to %q. Skipping volume size sync of PVCs.", c.OpConfig.StorageResizeMode)
		
```

File: pkg/cluster/sync.go [hunk 8]
```diff
@@ -1379,6 +1413,16 @@ func (c *Cluster) syncLogicalBackupJob() error {
...
+		if changed, _ := c.compareAnnotations(job.Annotations, desiredJob.Annotations); changed {
+			patchData, err := metaAnnotationsPatch(desiredJob.Annotations)
+			if err != nil {
+				return fmt.Errorf("could not form patch for the logical backup job %q: %v", jobName, err)
+			}
+			_, err = c.KubeClient.CronJobs(c.Namespace).Patch(context.TODO(), jobName, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+			if err != nil {
+				return fmt.Errorf("could not patch annotations of the logical backup job %q: %v", jobName, err)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (hughcapet) [reply pkg/cluster/sync.go:1425]: there it is done for pods

[the remark this answers is not in the document]
Comment (FxKu) [reply pkg/cluster/sync.go:1425]: got it

**Review by FxKu — APPROVED** [review]

**Comment by FxKu** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Pull request merged**